pkg/istio-agent/xds-proxy.go

```diff
@@ -63,37 +70,27 @@ type XdsProxy struct {
 
 var proxyLog = log.RegisterScope("xdsproxy", "XDS Proxy in Istio Agent", 0)
 
-// XDS Proxy proxies all XDS requests from envoy to istiod, in addition to allowing subsystems inside
-// the agent to also communicate with either istiod/envoy (eg dns, sds, etc).
-// The goal here is to consolidate all xds related connections to istiod/envoy into a
-// single tcp connection with multiple gRPC streams.
-// TODO: Right now, the workloadSDS server and gatewaySDS servers are still separate
-// connections. These need to be consolidated
-func initXdsProxy(sa *Agent, isSidecar bool) (*XdsProxy, error) {
+func initXdsProxy(ia *Agent) (*XdsProxy, error) {
 	var err error
 	proxy := &XdsProxy{
-		istiodAddress: sa.proxyConfig.DiscoveryAddress,
-		clusterID:     sa.secOpts.ClusterID,
+		istiodAddress:  ia.proxyConfig.DiscoveryAddress,
+		clusterID:      ia.secOpts.ClusterID,
+		localDNSServer: ia.localDNSServer,
 	}
 
 	if err = proxy.initDownstreamServer(); err != nil {
 		return nil, err
 	}
 
-	if proxy.istiodDialOptions, err = buildUpstreamClientDialOpts(sa); err != nil {
+	if proxy.istiodDialOptions, err = buildUpstreamClientDialOpts(ia); err != nil {
 		return nil, err
 	}
 
-	// we dont need dns server on gateways
-	if sa.cfg.DNSCapture && isSidecar {
-		if proxy.localDNSServer, err = dns.NewLocalDNSServer(sa.cfg.ProxyNamespace, sa.cfg.ProxyDomain); err != nil {
-			return nil, err
-		}
-		proxy.localDNSServer.StartDNS()
-	}
-
 	go func() {
-		_ = proxy.downstreamGrpcServer.Serve(proxy.downstreamListener)
+		if err = proxy.downstreamGrpcServer.Serve(proxy.downstreamListener); err != nil {
```

Comment:
isn't serve always going to terminate with an error after receiving kill ?

Reply:
Yeah..